include/ibamr/AdvDiffConservativeCUIConvectiveOperator.h

```diff
@@ -0,0 +1,221 @@
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_AdvDiffConservativeCUIConvectiveOperator
+#define included_IBAMR_AdvDiffConservativeCUIConvectiveOperator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibamr/config.h>
+
+#include "ibamr/ConvectiveOperator.h"
+#include "ibamr/ibamr_enums.h"
+
+#include "ibtk/ibtk_utilities.h"
+
+#include "CellVariable.h"
+#include "CoarsenAlgorithm.h"
+#include "FaceVariable.h"
+#include "IntVector.h"
+#include "PatchHierarchy.h"
+#include "RefineAlgorithm.h"
+#include "RefinePatchStrategy.h"
+#include "tbox/Database.h"
+#include "tbox/Pointer.h"
+
+#include <string>
+#include <vector>
+
+namespace SAMRAI
+{
+namespace solv
+{
+template <int DIM, class TYPE>
+class SAMRAIVectorReal;
+template <int DIM>
+class RobinBcCoefStrategy;
+} // namespace solv
+namespace xfer
+{
+template <int DIM>
+class CoarsenSchedule;
+template <int DIM>
+class RefineSchedule;
+} // namespace xfer
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class AdvDiffConservativeCUIConvectiveOperator is a concrete ConvectiveOperator
+ * that implements a upwind convective differencing operator based on the
+ * cubic upwind interpolation (CUI).
+ *
+ * Class AdvDiffConservativeCUIConvectiveOperator computes the convective derivative which is in
+ * the form of div(Q1*Q2*u) using the CUI method described by Waterson and Deconinck,
+ * and Patel and Natarajan.
+ *
+ *
+ * References
+ * Waterson, NP. and Deconinck, H., <A HREF="https://www.sciencedirect.com/science/article/pii/S002199910700040X">
+ * Design principles for bounded higher-order convection schemes – a unified approach</A>
+ *
+ * Patel, JK. and Natarajan, G., <A HREF="https://www.sciencedirect.com/science/article/pii/S0045793014004009">
+ * A generic framework for design of interface capturing schemes for multi-fluid flows</A>
+ *
+ * \see AdvDiffSemiImplicitHierarchyIntegrator
+ */
+class AdvDiffConservativeCUIConvectiveOperator : public ConvectiveOperator
+{
+public:
+    /*!
+     * \brief Class constructor.
+     */
+    AdvDiffConservativeCUIConvectiveOperator(std::string object_name,
+                                             SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q1_var,
+                                             SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q2_var,
+                                             SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                                             ConvectiveDifferencingType difference_form,
```

Comment:
Should this class always use "conservative" differencing?

Reply:
Following up from above --- if the class always uses conservative differencing, then it should not take in `difference_form` as an argument here.

Reply:
Yes. Now the class `AdvDiffCUIConservativeConvectiveOperator` do not take `difference_form` as an argument.